C/C++ targets which can be compiled/linked and used in python_dist() with ctypes

cosmicexplorer opened the pull request:
### Problem

It is currently possible to expose native code to Python by compiling it in a `python_dist()` target, specifying C or C++ source files as a `distutils.core.Extension` in the `setup.py` file, as well as in the target's sources. `python_dist()` was introduced in #5141. We introduced a "native toolchain" to compile native sources for this use case in #5490.

Exposing Python code this way requires using the Python native API and `#include <Python.h>` in your source files. However, python can also interact with native code that does not use the Python native API, using the provided `ctypes` library. For this to work, the `python_dist()` module using `ctypes` needs to have a platform-specific shared library provided within the package. This PR introduces the targets, tasks, and subsystems to compile and link a shared library from native code, then inserts it into the `python_dist()` where it is easily accessible.

### Solution

- Introduce the `ctypes_compatible_c_library()` target which covers C sources (`ctypes_compatible_cpp_library()` for C++), and can specify what to name the shared library created from linking the object files compiled from its sources and dependencies.
- Introduce `CCompile`, `CppCompile`, and `LinkSharedLibraries` to produce the shared libraries from the native sources. The compile tasks use options `CCompileSettings` or `CppCompileSettings` to define file extensions for "header" and "source" files.  
- Introduce the `CCompileSettings` and `CppCompileSettings` subsystems to control compile settings for those languages.
- Convert `BuildLocalPythonDistributions` to proxy to the native backend through the new `PythonNativeCode` subsystem.
- Move all the `BinaryTool` subsystems to a `subsystems/binaries` subdirectory, and expose them to the v2 engine through `@rule`s defined in the subsystem's file.
- Move some of the logic in `pex_build_util.py` to `setup_py.py`, and expose datatypes composing the setup.py environment through `@rule`s in `setup_py.py`. `SetupPyRunner.for_bdist_wheel()` was created to set the wheel's platform, if the `python_dist()` target contains any native sources of its own, or depends on any `ctypes_compatible_*_library`s. 

**Note:** the new targets are specifically prefixed with `ctypes_compatible_` because we don't yet eclipse the functionality of `contrib/cpp`. When the targets become usable for more than this one use case, the name should be changed.

### Result

To see how to link up native and Python code with `ctypes`, here's (most of) the contents of `testprojects/src/python/python_distribution/ctypes`:
*BUILD*:
```python
ctypes_compatible_c_library(
  name='c_library',
  sources=['some_math.h', 'some_math.c', 'src-subdir/add_three.h', 'src-subdir/add_three.c'],
  ctypes_dylib=native_artifact(lib_name='asdf-c'),
)

ctypes_compatible_cpp_library(
  name='cpp_library',
  sources=['some_more_math.hpp', 'some_more_math.cpp'],
  ctypes_dylib=native_artifact(lib_name='asdf-cpp'),
)

python_dist(
  sources=[
    'setup.py',
    'ctypes_python_pkg/__init__.py',
    'ctypes_python_pkg/ctypes_wrapper.py',
  ],
  dependencies=[
    ':c_library',
    ':cpp_library',
  ],
)
```
*setup.py*:
```python
setup(
  name='ctypes_test',
  version='0.0.1',
  packages=find_packages(),
  # Declare two files at the top-level directory (denoted by '').
  data_files=[('', ['libasdf-c.so', 'libasdf-cpp.so'])],
)
```
*ctypes_python_pkg/ctypes_wrapper.py*:
```python
import ctypes
import os


def get_generated_shared_lib(lib_name):
  # These are the same filenames as in setup.py.
  filename = 'lib{}.so'.format(lib_name)
  # The data files are in the root directory, but we are in ctypes_python_pkg/.
  rel_path = os.path.join(os.path.dirname(__file__), '..', filename)
  return os.path.normpath(rel_path)


asdf_c_lib_path = get_generated_shared_lib('asdf-c')
asdf_cpp_lib_path = get_generated_shared_lib('asdf-cpp')

asdf_c_lib = ctypes.CDLL(asdf_c_lib_path)
asdf_cpp_lib = ctypes.CDLL(asdf_cpp_lib_path)

def f(x):
  added = asdf_c_lib.add_three(x)
  multiplied = asdf_cpp_lib.multiply_by_three(added)
  return multiplied
```

Now, the target `testprojects/src/python/python_distribution/ctypes` can be depended on in a BUILD file, and other python code can freely use `from ctypes_python_pkg.ctypes_wrapper import f` to start jumping into native code.

#### Follow-up issues:
1. #5933 
2. #5934 
3. #5949
4. #5950
5. #5951
6. #5962
7. #5967
8. #5977

cosmicexplorer commented on src/python/pants/backend/native/targets/native_library.py:
```diff
@@ -0,0 +1,65 @@
+
+    logger.debug("sources_field.sources: {}".format(sources_field.sources))
+
+    if provides and not isinstance(provides, NativeArtifact):
```
Yes, `provides` is the exact opposite of the right term here. I think the right interface is likely still as a kwarg of the target as opposed to a new wrapper target -- potentially `exports=` would be appropriate?

stuhood replied:
The name `exports` is overloaded as well (see the usage on jvm targets).

Is this choosing an importable name for the library? ie `lib$blah`?

cosmicexplorer replied:
`native_artifact()` is a descriptor which currently is used to form the name of the shared library which is then consumed in a `python_dist()` that depends on the c/c++ target (the shared library is named `lib<name>.so`). The idea is that `native_artifact()` does not imply specifically a dynamic or static library, which is an implementation detail.

cosmicexplorer commented on src/python/pants/backend/native/subsystems/native_toolchain.py:
```diff
@@ -41,61 +42,72 @@ class NativeToolchain(Subsystem, ExecutablePathProvider):
+  #   'linux': lambda: Get(Linker, Binutils, native_toolchain._binutils),
+  # })
+  if platform.normalized_os_name == 'darwin':
+    # TODO(cosmicexplorer): turn this into LLVM when lld works.
```
Yes! It was #5663. Noted in the comment.

cosmicexplorer commented on src/python/pants/backend/native/tasks/c_compile.py:
```diff
@@ -0,0 +1,116 @@
+    return self._request_single(CCompiler, self._toolchain)
+
+  # FIXME: note somewhere that this means source file names within a target must be unique (even if
+  # the files are in different subdirectories) -- check this at the target level!!!
```
Removed this comment, as we already explode with a detailed error message when we invoke `NativeCompile#get_sources_headers_for_target()`.

cosmicexplorer commented on src/python/pants/backend/native/tasks/c_compile.py:
```diff
@@ -0,0 +1,116 @@
+      return ObjectFiles(output_dir, [])
+
+
+    # TODO: add -fPIC, but only to object files used for shared libs (how do we determine that?) --
```
Added a link to #5788 in the comment.

cosmicexplorer commented on src/python/pants/backend/python/tasks/pex_build_util.py:
```diff
@@ -52,7 +54,13 @@ def is_python_binary(tgt):
 def tgt_closure_has_native_sources(tgts):
   """Determine if any target in the current target closure has native (c or cpp) sources."""
-  return any(tgt.has_native_sources for tgt in tgts)
+  # FIXME: introduce source_target_constraint and rely on it for all tasks -- override
```
I had the same thought immediately as I wrote it! Removed.

cosmicexplorer commented on src/python/pants/backend/python/tasks/setup_py.py:
```diff
@@ -79,21 +95,87 @@ def _setup_command(self):
+
+@rule(SetupPyNativeTools, [Select(NativeToolchain)])
+def get_setup_py_native_tools(native_toolchain):
+  c_compiler = yield Get(CCompiler, NativeToolchain, native_toolchain)
```
A fantastic comment! I have made the change.

cosmicexplorer commented on src/python/pants/backend/native/tasks/native_compile.py:
```diff
@@ -0,0 +1,176 @@
+  # TODO: document how these two are supposed to both produce an ObjectFiles (getting from the cache
+  # vs getting from a compile).
+  @abstractmethod
+  def collect_cached_objects(self, versioned_target):
```
This was a great comment! I have taken your advice, and in doing so converted the interface to be two abstract methods: `get_compiler(self)` and `compile(self, compile_request)`, and moving much of the logic in both `NativeTask` as well as `C/CppCompile` into `NativeCompile` instead (including generating the compile request). This also makes the `execute(self)` method in `NativeCompile` much easier to follow.

To move the `strict_deps` logic out of `NativeTask`, I introduced a separate `NativeTargetDependencies` product to cache the target's dependencies, which the `LinkSharedLibraries` task consumes (with a note that it should probably store something besides just the target).

I also moved all the options to the new `C/CppCompileSettings` subsystems, which took a lot of unnecessary complexity out of the tasks as well.

cosmicexplorer commented on src/python/pants/backend/native/tasks/native_task.py:
```diff
@@ -0,0 +1,69 @@
+  def register_options(cls, register):
+    super(NativeTask, cls).register_options(register)
+
+    register('--strict-deps', type=bool, default=True, fingerprint=True, advanced=True,
```
Done, see above comment.

cosmicexplorer commented on src/python/pants/backend/native/tasks/native_compile.py:
```diff
@@ -0,0 +1,176 @@
+  @classmethod
+  def register_options(cls, register):
+    super(NativeCompile, cls).register_options(register)
+
```
I have elected to move all of the options into subsystems, see the below comment.

cosmicexplorer commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -53,80 +58,87 @@ def implementation_version(cls):
+    native_toolchain = NativeToolchain.scoped_instance(self)
+    return self._request_single(SetupPyNativeTools, native_toolchain)
+
+  # TODO: This should probably be made into a class property, when that is made.
```
Linked to my ticket!

cosmicexplorer commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -53,80 +58,87 @@ def implementation_version(cls):
-  def _native_toolchain_instance(self):
-    return NativeToolchain.scoped_instance(self)
+  def _request_single(self, product, subject):
+    # FIXME(#4769): This is not supposed to be exposed to Tasks yet -- see #4769 to track the status
```
Left a TODO to do that!

cosmicexplorer commented on src/python/pants/backend/python/targets/python_distribution.py:
```diff
@@ -19,6 +19,8 @@ class PythonDistribution(Target):
 
   default_sources_globs = '*.py'
 
+  native_source_extensions = ['.c', '.cpp', '.cc']
```
I have moved these into the new `PythonNativeCode` subsystem, which is now the bridge to the `NativeToolchain` for `BuildLocalPythonDistributions`.

cosmicexplorer commented:
I have updated the `python_dist()` testing to incorporate the changes, and moved all the compile options into new subsystems. I have also introduced the `PythonNativeCode` subsystem to connect the native backend to the python backend. I cleaned up comments, docstrings, and made TODOs and FIXMEs point to a ticket or rewrote them to be self-explanatory.

I will now try to see if there's a natural point this can be broken off at. This really should have unit tests for `CCompile`, `CppCompile`, and `LinkSharedLibraries`, but this is large enough as it is, so I will make a followup ticket for this.

cosmicexplorer commented:
Left a FIXME pointing to where #5662 can jump in immediately.

cosmicexplorer commented:
CI is mostly green on this except for the one integration test shard which timed out -- I'll figure it out after I try to sleep if that ends up not being spurious. I would really love (1) review of the code here (which is basically an entire language pipeline) (2) guidance on how to make this easiest to merge -- I'm very open to splitting chunks off into separate diffs and/or making detailed followup issues. Thanks a lot in advance for any help.

cosmicexplorer commented:
@benjyw I have just squashed it into 5 artfully arranged commits -- let me know if there's more context I can provide.

cosmicexplorer commented:
On python integration test shard 3, Travis keeps reporting as a timeout, but there's actually a test failure that I seem to be able to repro locally? I can't figure out what might be happening from the error message, and I don't know if or where zinc keeps any logs that might have more info.

```bash
> ./pants test tests/python/pants_test/backend/project_info/tasks:export_integration -- -k test_distributions_and_platforms
# ...
 tests/python/pants_test/backend/project_info/tasks/test_export_integration.py F [100%]
 
 ==================== FAILURES ====================
  ExportIntegrationTest.test_distributions_and_platforms 
 
 self = <pants_test.backend.project_info.tasks.test_export_integration.ExportIntegrationTest testMethod=test_distributions_and_platforms>
 args = (), kwargs = {}, env_var_value = 'ivy'
 
     def wrapper(self, *args, **kwargs):
       for env_var_value in ('ivy', 'coursier'):
         with environment_as(HERMETIC_ENV='PANTS_RESOLVER_RESOLVER', PANTS_RESOLVER_RESOLVER=env_var_value):
 >         f(self, *args, **kwargs)
 
 .pants.d/pyprep/sources/01f4e3f978c89e2bb0e1e36c597cac855ba92cae/pants_test/pants_run_integration_test.py:68: 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 .pants.d/pyprep/sources/01f4e3f978c89e2bb0e1e36c597cac855ba92cae/pants_test/backend/project_info/tasks/test_export_integration.py:188: in test_distributions_and_platforms
     '--jvm-distributions-paths={'
 .pants.d/pyprep/sources/01f4e3f978c89e2bb0e1e36c597cac855ba92cae/pants_test/backend/project_info/tasks/test_export_integration.py:46: in run_export
     self.assert_success(pants_run)
 .pants.d/pyprep/sources/01f4e3f978c89e2bb0e1e36c597cac855ba92cae/pants_test/pants_run_integration_test.py:392: in assert_success
     self.assert_result(pants_run, self.PANTS_SUCCESS_CODE, expected=True, msg=msg)
 .pants.d/pyprep/sources/01f4e3f978c89e2bb0e1e36c597cac855ba92cae/pants_test/pants_run_integration_test.py:413: in assert_result
     assertion(value, pants_run.returncode, error_msg)
 E   AssertionError: /home/cosmicexplorer/tools/pants/pants --no-pantsrc --pants-workdir=/home/cosmicexplorer/tools/pants/.pants.d/tmp/tmpjhgG7P.pants.d --kill-nailguns --print-exception-stacktrace=True export --output-file=/home/cosmicexplorer/tools/pants/.pants.d/tmp/tmpjhgG7P.pants.d/export_out.txt examples/src/java/org/pantsbuild/example/hello/simple --no-export-libraries --jvm-platform-default-platform=java7 --jvm-platform-platforms={ "java7": {"source": "1.7", "target": "1.7", "args": [ "-X123" ]}, "java8": {"source": "1.8", "target": "1.8", "args": [ "-X456" ]}} --jvm-distributions-paths={ "macos": [ "/Library/JDK" ], "linux": [ "/usr/lib/jdk7", "/usr/lib/jdk8"]}
 E   returncode: 1
 E   stdout:
 E   
 E          FAILURE: Compilation failure: Failed jobs: compile(examples/src/java/org/pantsbuild/example/hello/simple:simple)
 E   
 E   stderr:
 E          .....
 E          compile(examples/src/java/org/pantsbuild/example/hello/simple:simple) failed: Zinc compile failed.INFO] killing nailgun server pid=15848
 E   
 E          FAILURE
                      generated xml file: /home/cosmicexplorer/tools/pants/.pants.d/test/pytest/tests.python.pants_test.backend.project_info.tasks.export_integration/junitxml/TEST-tests.python.pants_test.backend.project_info.tasks.export_integration.xml
# ...
```

stuhood commented on src/python/pants/backend/native/subsystems/native_compile_settings.py:
```diff
@@ -0,0 +1,54 @@
+             help='The default for the "fatal_warnings" argument for targets of this language.')
+
+    # TODO: make a list of file extension option type?
+    register('--header-file-extensions', type=list, default=cls.default_header_file_extensions,
```
It feels a bit premature to make the "file extensions" into options. Certainly it would take subclassing to define a new compiler, in which case having them in options would be redundant.

cosmicexplorer replied:
I can understand the idea here, but the compile task will error out if some source files don't have the right extensions -- this is so that we can differentiate headers and sources. Not making them options means users have a restricted set of file names allowed for header and source files unless they make their own subclass in a plugin (and porting a project to Pants might require some work). The alternative is to have separate `headers=` and `sources=` fields, which is very doable -- I just thought this was a cleaner way to obtain headers and sources, and not restrict file names, and removes some (unnecessary?) repetitive typing for the writer of the BUILD file compared to the alternative of splitting `headers=` and `sources=` into separate fields.

stuhood commented on src/python/pants/backend/native/subsystems/native_compile_settings.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class CppCompileSettings(NativeCompileSettings):
+  options_scope = 'cpp-compile-settings'
```
Including the word `settings` in the scope name is redundant, since all scopes refer to settings. Would just do `c-compile` and `cpp-compile`.

cosmicexplorer replied:
Done!

stuhood commented on src/python/pants/backend/native/targets/native_library.py:
```diff
@@ -0,0 +1,78 @@
+  """A class wrapping targets containing sources for C-family languages and related code."""
+
+  @classmethod
+  def produces_ctypes_dylib(cls, target):
```
`ctypes` is python specific, so it's a bit odd for references to it to be here.

Presumably you're trying to avoid needing to create new wrapper target types for each language that might consume the library... if so, it might be worth discouraging the next person who comes along from doing that (and instead encouraging them to create the separate target to have a relationship like `python_library depends on ctype_library dependes on c_library`). Alternatively, consider just doing that here?

cosmicexplorer replied:
The reason for adding `ctypes` to any symbol names was so that it's clear when used they are not capable of any other use yet (and also, because of name clashes with `contrib/cpp`). The idea is that there is just `c/cpp_library`, declaring maybe a `produces_library=native_artifact()`, and that can be a dependency for multiple different languages, and the names would be modified as functionality is added.

stuhood commented on src/python/pants/backend/native/targets/native_library.py:
```diff
@@ -0,0 +1,78 @@
+from pants.build_graph.target import Target
+
+
+class NativeLibrary(Target):
```
Is this supposed to be abstract?

stuhood commented on src/python/pants/backend/native/targets/native_library.py:
```diff
@@ -0,0 +1,78 @@
+
+  @classmethod
+  def alias(cls):
+    return 'ctypes_compatible_cpp_library'
```
Again, ctypes references are odd.

cosmicexplorer replied:
See my other commment [above](https://github.com/pantsbuild/pants/pull/5815#discussion_r194574698).

stuhood commented on src/python/pants/backend/native/tasks/c_compile.py:
```diff
@@ -0,0 +1,89 @@
+          .format(cmd, compile_request, e),
+          e)
+
+      rc = process.wait()
```
Calling wait outside of `try:except` seem odd. Does `wait` never return exceptions?

stuhood commented on src/python/pants/backend/native/tasks/cpp_compile.py:
```diff
@@ -0,0 +1,87 @@
+      '-I{}'.format(os.path.abspath(inc_dir)) for inc_dir in compile_request.include_dirs
+    ] + [os.path.abspath(src) for src in sources]
+
+    with self.context.new_workunit(name='cpp-compile', labels=[WorkUnitLabel.COMPILER]) as workunit:
```
The two subclasses of `NativeCompile` seem to have a lot of redundancy around process spawning. It would be good to remove that if possible.

cosmicexplorer replied:
Done, making `c_compile.py` and `cpp_compile.py` significantly smaller and easier to understand. Also ended up breaking out #5951 with a FIXME.

jsirois commented:
@cosmicexplorer have you tried running the pants command the test runs directly? Looks like this approximates it:

```bash
#!/usr/bin/env bash

TEST_TMP_DIR=/tmp/test

rm -rf "${TEST_TMP_DIR}"
mkdir -p "${TEST_TMP_DIR}"

./pants \
        --no-pantsrc \
        --pants-workdir="${TEST_TMP_DIR}/.pants.d" \
        --kill-nailguns \
        --print-exception-stacktrace=True \
        export \
                --output-file="${TEST_TMP_DIR}/export_out.txt" \
                examples/src/java/org/pantsbuild/example/hello/simple \
                --no-export-libraries \
                --jvm-platform-default-platform=java7 \
                --jvm-platform-platforms='{ "java7": {"source": "1.7", "target": "1.7", "args": [ "-X123" ]}, "java8": {"source": "1.8", "target": "1.8", "args": [ "-X456" ]}}' \
                --jvm-distributions-paths='{ "macos": [ "/Library/JDK" ], "linux": [ "/usr/lib/jdk7", "/usr/lib/jdk8"]}'
```

stuhood commented on src/python/pants/backend/native/tasks/native_compile.py:
```diff
@@ -0,0 +1,248 @@
+])): pass
+
+
+# TODO: verify that filenames are valid fileNAMES and not deeper paths? does this matter?
```
Would likely be better to use `Snapshot`s moving forward, which handles this for you. But that's a followup.

cosmicexplorer replied:
Added link to #5950.

stuhood commented on src/python/pants/backend/python/subsystems/python_native_code.py:
```diff
@@ -0,0 +1,123 @@
+
+  options_scope = 'python-native-code'
+
+  default_native_source_extensions = ['.c', '.cpp', '.cc']
```
It's a bit concerning that this is redundant with the `cpp-compile` and `c-compile` subsystems... are different tools being invoked that do not all use the same extensions? Would it be reasonable to just default to supporting the intersection of these two sets, and then configure it on the relevant language subsystem?

cosmicexplorer replied:
This shouldn't be here at all, but checking if a `python_dist()` target has "native" sources has been done by checking whether it has sources by any of these extensions. What it should really be doing is scanning the requirements closure to see if any of them have platform-specific code. That somehow didn't get made into a followup issue, which I will do now.

cosmicexplorer replied:
Made into #5949.

stuhood commented on src/python/pants/backend/python/subsystems/python_native_code.py:
```diff
@@ -0,0 +1,123 @@
+
+    register('--native-source-extensions', type=list, default=cls.default_native_source_extensions,
+             fingerprint=True, advanced=True,
+             help='The extensions recognized for native source files in `python_dist()` sources.')
```
ditto option feeling like overkill

cosmicexplorer replied:
See #5949.

stuhood commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -44,89 +45,104 @@ def product_types(cls):
-            self.context.build_graph.inject_dependency(dependent, req_lib_addr)
-          local_wheel_products.add(vt.target, os.path.dirname(dist)).append(os.path.basename(dist))
+    dist_targets = self.context.targets(is_local_python_dist)
+    if not dist_targets:
```
A reminder that this type of edit (dedenting a bunch of stuff) makes the diff useless... best to make this kind of change in reviews where relatively little else is going on.

cosmicexplorer replied:
Reverted! Will remember for the future.

stuhood commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -135,45 +151,88 @@ def _copy_sources(self, dist_tgt, dist_target_dir):
+      basename = os.path.basename(shared_lib.path)
+      # NB: We convert everything to .so here so that the setup.py can just
+      # declare .so to build for either platform.
+      resolved_outname = re.sub(r'\..*\Z', '.so', basename)
```
IIRC, `.so` is platform specific. Would this be `.dylib` on OSX? Regardless, make sure OSX coverage via #5920.

cosmicexplorer replied:
The python code is assumed to only interact with the shared library through `ctypes`, which does not require the shared library to end in `.dylib` on osx when loading it by file path. As in the comment, this avoids platform-specific code in getting the path to the shared library to declare in the `setup.py` or in the python source.

cosmicexplorer replied:
Will add the ctypes integration testing to the osx shard.

stuhood commented on src/python/pants/util/collections.py:
```diff
@@ -19,3 +19,10 @@ def recursively_update(d, d2):
+def assert_single_element(iterable):
+  """Convert `iterable` to a list, assert that the list has one element, then return the element."""
+  result = list(iterable)
+  assert(len(result) == 1)
```
Using an `assert` for something like this isn't that friendly. Should probably actually have an error message.

cosmicexplorer replied:
Fixed in 3be53f3!

stuhood commented on src/python/pants/util/dirutil.py:
```diff
@@ -500,3 +500,10 @@ def split_basename_and_dirname(path):
   return (os.path.dirname(path), os.path.basename(path))
+
+
+def narrow_relative_paths(cur_root_dir, new_root_subdir, rel_paths):
```
This is pretty specific... would not add it here unless you've used the pattern multiple times.

cosmicexplorer replied:
Reverted in 97787f5!

cosmicexplorer commented:
@jsirois that's very helpful and I'll keep that in mind for the future! Running that script produced the same output in my above comment, however, in `/tmp/test/.pants.d/run-tracker/pants_run_2018_06_11_17_13_23_313_bf5b63de9081406197dbc9a50fd108ff/tool_outputs/zinc-96d4bd3d-0c0f-4f1c-adef-4c4cc0359f7a.stdout`, I found:
```
[error] Unknown option: -X123
[error] See zinc-compiler -help for information about options
```

I realized that this error is likely due to some of the changes I made at the beginning extending `DependencyContext`. Will look into fixing this now.

CMLivingston commented on src/python/pants/backend/python/subsystems/python_native_code.py:
```diff
@@ -0,0 +1,125 @@
+    platform_names = platforms_with_sources.keys()
+
+    # There will always be at least 1 platform, because we checked that they have native sources.
+    assert(len(platform_names) >= 1)
```
Would be great to add a string here explaining the assertion error.

cosmicexplorer replied:
Made it into an exception!

CMLivingston commented on tests/python/pants_test/backend/python/tasks/test_ctypes_integration.py:
```diff
@@ -0,0 +1,71 @@
+      self.assert_success(pants_run)
+
+      # Check for the pex and for the wheel produced for our python_dist().
+      pex = os.path.join(tmp_dir, 'bin.pex')
```
Would be good to wrap this pants run in a try/except and clean up the output pex after the test finished (successfully or otherwise).

cosmicexplorer replied:
I think that's handled by the `temporary_dir()` contextmanager? Because we use that for the `dist/` directory. Is something staying behind when you run the test?

CMLivingston commented on src/python/pants/backend/native/targets/native_library.py:
```diff
@@ -0,0 +1,79 @@
+    sources_field = self.create_sources_field(sources, address.spec_path, key_arg='sources')
+    payload.add_fields({
+      'sources': sources_field,
+      'ctypes_dylib': ctypes_dylib,
```
This naming usage of `dylib` is slightly confusing as it clashes with the osx-specific file extension. Is this osx specific at all? Would it improve readability to something like `ctypes_dynamic_library` or the like?

cosmicexplorer replied:
*[see another comment about naming here](https://github.com/pantsbuild/pants/pull/5815#discussion_r194590807)*

This isn't osx-specific, it's supposed to state "we can export the contents of this library into an implementation-defined format (e.g. static *or* dynamic lib) with a native ABI to another language". The use of `ctypes_*` everywhere is so it's clear what the capabilities are right now, and is basically an explicit choice to *not* make a final decision on what to call things right now. I especially don't want to adopt names that imply things like static or dynamic libraries at the target level.

I agree that for those reasons and for your reasons, the name `ctypes_dylib` is confusing. Do you think `ctypes_native_library=` might make sense?

CMLivingston commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -44,89 +45,103 @@ def product_types(cls):
+          reqs_to_resolve.add(req)
 
-    if not reqs_to_resolve:
+    if len(reqs_to_resolve) == 0:
```
Why switch this?

cosmicexplorer replied:
No reason, reverted. Thanks, and please point out more issues like this, it's not intentional and I should have found this when going over the diff.

CMLivingston commented on src/python/pants/backend/python/tasks/setup_py.py:
```diff
@@ -79,21 +95,87 @@ def _setup_command(self):
+  }
+
+  # The `python_dist` target builds for the current platform only.
+  # FIXME: why does it build for the current platform only?
```
This was a cheap catch-all for preventing errors when the closure has platform-specific sources or dependencies.

cosmicexplorer replied:
Could you describe this a bit more? Why are we indexing with `['current']` here? Or is there more info in #5825 somewhere?

CMLivingston commented on src/python/pants/backend/python/tasks/build_local_python_distributions.py:
```diff
@@ -44,89 +45,103 @@ def product_types(cls):
 
+      with self.invalidated(dist_targets, invalidate_dependents=True) as invalidation_check:
         for vt in invalidation_check.invalid_vts:
-          if vt.target.dependencies:
```
It is worth mentioning that allowing dependencies here breaks the original vision for python_dist as a Pants-specific apparatus for depending on a properly-packaged and readily-consumable python distributions (that exists in the python ecosystem with nothing Pants-specific at all). Clearly it is crucial to this PR that we allow dependencies and I don't think it it is a problem, but I wanted to mention that we may be stepping on/adding additional conceptual loading onto what a "python distribution" is in Pants context. It may make sense to rename the target altogether or add some explicit docs/visibility to the fact that this target now offers more native integrations into Pants-land (via dependencies) that the original `python_dist` would not allow.

cosmicexplorer replied:
I totally see this! I guess my thought was that: this feature doesn't seem to stop us from supporting a "properly-packaged and readily-consumable python distribution" (and that was part of the rationale for why it works like this). Because dependencies aren't allowed in a "normal" `python_dist()`, and because the only intrusion we have into the `setup.py` is adding `data_files=[('', ['libasdf-c.so', 'libasdf-cpp.so'])],` which seems like verbatim what you would add if you were doing externally-compiled artifacts without using Pants. The fact that it ends up setting the platform might be a little magic (and needs documentation), but again it also seems like what you would do if your goal was to externally compile some native artifacts and insert them into your `python_dist()`. This technique works *seamlessly* with native code in the `python_dist()` itself, because we're just adding a resource. Could you explain what concepts we're stepping on here? It's not clear to me.

If you mean "make it more clear the tradeoffs of this technique vs compiling native code in the `python_dist()`", I'm 100% with you, there are distinct use cases for both (using the Python C/C++ API vs accessing native code with `ctypes`), and that deserves a clear set of documentation. If there are other inconsistencies that should be clearly noted in a page on pantsbuild.org, let me know, I'll make a followup issue for that now.

cosmicexplorer commented:
> it should integrate seamlessly with these changes

@CMLivingston Really glad to hear that! I left a few responses to your comments -- I'm totally open to changing the target's name or almost anything else.

benjyw commented:
I'll be honest, this change is so massive it's hard to wrap my brain around for a detailed review. Overall it looks fine, structurally. It's probably better to get this in more or less as-is and follow up later with any tweaks and fixes. Is there anything specific you want me to drill down on?

cosmicexplorer commented:
@benjyw Thanks for being frank about your ability to review this! I think it would be useful to have your input on:

- What do you think about the `Platform` datatype in `src/python/pants/backend/native/config/environment.py`, and the method `resolve_platform_specific()` (and its usages)? Is this method of implementing platform-specific code going to become annoying at any point?
- For any use case you personally might have (and/or any use cases you happen to be aware of), what features would you expect and/or demand in pants as we deepen support for C and C++ (both big and small)?

stuhood commented:
@cosmicexplorer : Would you mind opening an issue about deprecating the C/C++ contrib module? I don't think we need to provide a "path" to this review, since I think it was effectively unused.

stuhood commented:
Hm. And actually... would it be possible to move the c/cpp tasks into contrib? I think so, since python only depends on c/cpp via products?

benjyw commented on src/python/pants/backend/native/config/environment.py:
```diff
@@ -0,0 +1,84 @@
+
+  _NORMALIZED_OS_NAMES = frozenset(all_normalized_os_names())
+
+  def resolve_platform_specific(self, platform_specific_funs):
```
Should probably document this method, specifically what shape the argument should take.

benjyw replied:
And the shape of the return value of course.

cosmicexplorer replied:
Will do! #5933 may eventually turn into an alternative implementation of platform-specific logic, pending developments in creating subsystems and/or using variants in the v2 engine (neither of which I am currently familiar with), so making it clear why we're doing it this way now seems especially important.

cosmicexplorer commented:
> would it be possible to move the c/cpp tasks into contrib? I think so, since python only depends on c/cpp via products?

Yes! I will do that now, I believe it should be very straightforward.

cosmicexplorer commented:
Created #5970 to move the native backend into contrib!
